src/driver/driver.cpp

```diff
@@ -171,22 +173,42 @@ void EvolutionDriver::OutputCycleDiagnostics() {
   const int dt_precision = std::numeric_limits<Real>::max_digits10 - 1;
   if (tm.ncycle_out != 0) {
     if (tm.ncycle % tm.ncycle_out == 0) {
-      if (Globals::my_rank == 0) {
-        std::uint64_t zonecycles =
-            (pmesh->mbcnt - mbcnt_prev) *
-            static_cast<std::uint64_t>(pmesh->GetNumberOfMeshBlockCells());
-        std::cout << "cycle=" << tm.ncycle << std::scientific
-                  << std::setprecision(dt_precision) << " time=" << tm.time
-                  << " dt=" << tm.dt << std::setprecision(2) << " zone-cycles/wsec = "
-                  << static_cast<double>(zonecycles) / timer_cycle.seconds();
-        // insert more diagnostics here
-        std::cout << std::endl;
-
-        // reset cycle related counters
-        timer_cycle.reset();
-        // need to cache number of MeshBlocks as AMR/load balance change it
-        mbcnt_prev = pmesh->mbcnt;
+      std::uint64_t zonecycles =
+          (pmesh->mbcnt - mbcnt_prev) *
+          static_cast<std::uint64_t>(pmesh->GetNumberOfMeshBlockCells());
+      const auto time_cycle_all = timer_cycle.seconds();
+      const auto time_cycle_step = time_cycle_all - time_LBandAMR;
+      std::cout << "cycle=" << tm.ncycle << std::scientific
+                << std::setprecision(dt_precision) << " time=" << tm.time
+                << " dt=" << tm.dt << std::setprecision(2) << " zone-cycles/wsec_step="
+                << static_cast<double>(zonecycles) / time_cycle_step
+                << " wsec_step=" << time_cycle_step;
+
+      // In principle load balancing based on a cost list can happens for non-AMR runs.
+      // TODO(future me) fix this when this becomes important.
+      if (pmesh->adaptive) {
+        std::cout << " zone-cycles/wsec="
+                  << static_cast<double>(zonecycles) / (time_cycle_step + time_LBandAMR)
+                  << " wsec_AMR=" << time_LBandAMR;
       }
+
+      // insert more diagnostics here
+      std::cout << std::endl;
+
+      // reset cycle related counters
+      timer_cycle.reset();
+      time_LBandAMR = 0.0;
+      // need to cache number of MeshBlocks as AMR/load balance change it
+      mbcnt_prev = pmesh->mbcnt;
+    }
+  }
+  if (tm.ncycle_out_mesh != 0) {
+    // output in fixed intervals
+    if (tm.ncycle % tm.ncycle_out_mesh == 0) {
+      pmesh->OutputMeshStructure(-1, false);
+      // output after mesh refinement (enabled by use of negative cycle number)
+    } else if (tm.ncycle_out_mesh < 0 && pmesh->modified) {
+      pmesh->OutputMeshStructure(-1, false);
```

Comment:
It might be nice to output something like
```
-------------------------------------------------------
// mesh structure here
-------------------------------------------------------
```
just to set aside the mesh structure from the regular output a little more and make it more readable.

Reply:
How does this look like?
```
cycle=28 time=5.1406510065595977e-05 dt=1.6932658600527559e-06 zone-cycles/wsec_step=2.24e+06 wsec_step=2.20e-01 zone-cycles/wsec=2.24e+06 wsec_AMR=1.54e-04
cycle=29 time=5.3099775925648732e-05 dt=1.6991523280397159e-06 zone-cycles/wsec_step=2.21e+06 wsec_step=2.22e-01 zone-cycles/wsec=2.21e+06 wsec_AMR=3.38e-04
cycle=30 time=5.4798928253688446e-05 dt=1.7045121209625266e-06 zone-cycles/wsec_step=2.20e+06 wsec_step=2.23e-01 zone-cycles/wsec=2.19e+06 wsec_AMR=1.72e-03
---------------------- Current Mesh structure ----------------------
Root grid = 8 x 8 x 8 MeshBlocks
Total number of MeshBlocks = 960
Number of physical refinement levels = 2
Number of logical  refinement levels = 5
  Physical level = 0 (logical level = 3): 480 MeshBlocks, cost = 480
  Physical level = 1 (logical level = 4): 224 MeshBlocks, cost = 224
  Physical level = 2 (logical level = 5): 256 MeshBlocks, cost = 256
--------------------------------------------------------------------
cycle=31 time=5.6503440374650975e-05 dt=1.7095544528615823e-06 zone-cycles/wsec_step=2.15e+06 wsec_step=2.28e-01 zone-cycles/wsec=2.15e+06 wsec_AMR=1.81e-04
cycle=32 time=5.8212994827512555e-05 dt=1.7143245852434286e-06 zone-cycles/wsec_step=2.24e+06 wsec_step=2.20e-01 zone-cycles/wsec=2.22e+06 wsec_AMR=1.27e-03
cycle=33 time=5.9927319412755980e-05 dt=1.7132534065589339e-06 zone-cycles/wsec_step=2.21e+06 wsec_step=2.22e-01 zone-cycles/wsec=2.21e+06 wsec_AMR=1.54e-04
cycle=34 time=6.1640572819314916e-05 dt=1.7123032206621521e-06 zone-cycles/wsec_step=2.12e+06 wsec_step=2.32e-01 zone-cycles/wsec=1.41e+06 wsec_AMR=1.16e-01
-------------- New Mesh structure after (de)refinement -------------
Root grid = 8 x 8 x 8 MeshBlocks
Total number of MeshBlocks = 1296
Number of physical refinement levels = 2
Number of logical  refinement levels = 5
  Physical level = 0 (logical level = 3): 456 MeshBlocks, cost = 456
  Physical level = 1 (logical level = 4): 392 MeshBlocks, cost = 392
  Physical level = 2 (logical level = 5): 448 MeshBlocks, cost = 448
--------------------------------------------------------------------
cycle=35 time=6.3352876039977064e-05 dt=1.7124939092645730e-06 zone-cycles/wsec_step=2.04e+06 wsec_step=3.25e-01 zone-cycles/wsec=2.04e+06 wsec_AMR=2.77e-04
cycle=36 time=6.5065369949241639e-05 dt=1.7139091408957631e-06 zone-cycles/wsec_step=2.15e+06 wsec_step=3.09e-01 zone-cycles/wsec=2.09e+06 wsec_AMR=9.22e-03
```

Reply:
That looks fantastic. :D